examples/cogvideo/train_cogvideox_lora.py

```diff
@@ -0,0 +1,1494 @@
+
+        for step, batch in enumerate(train_dataloader):
+            models_to_accumulate = [transformer]
+            if args.train_text_encoder:
+                models_to_accumulate.extend([text_encoder])
+
+            with accelerator.accumulate(models_to_accumulate):
+                videos = batch["videos"].to(dtype=vae.dtype)
+                prompts = batch["prompts"]
+
+                # encode prompts
+                prompt_embeds = compute_prompt_embeddings(
+                    tokenizer,
+                    text_encoder,
+                    prompts,
+                    accelerator.device,
+                    weight_dtype,
+                    requires_grad=args.train_text_encoder,
+                )
+
+                # Convert videos to latents
+                videos = videos.permute(0, 2, 1, 3, 4)  # [B, C, F, H, W]
+                model_input = vae.encode(videos).latent_dist.sample() * vae.config.scaling_factor
+                model_input = model_input.permute(0, 2, 1, 3, 4).to(dtype=weight_dtype)  # [B, F, C, H, W]
+
+                # Sample noise that will be added to the latents
+                noise = torch.rand_like(model_input)
+                batch_size, num_frames, num_channels, height, width = model_input.shape
+
+                # Sample a random timestep for each image
+                timesteps = torch.randint(
+                    0, scheduler.config.num_train_timesteps, (batch_size,), device=model_input.device
+                )
+                timesteps = timesteps.long()
+
+                # Prepare rotary embeds
+                image_rotary_emb = (
+                    prepare_rotary_positional_embeddings(
+                        height=args.height,
+                        width=args.width,
+                        num_frames=num_frames,
+                        vae_scale_factor_spatial=vae_scale_factor_spatial,
+                        patch_size=transformer.config.patch_size,
+                        attention_head_dim=transformer.config.attention_head_dim,
+                        device=accelerator.device,
+                    )
+                    if transformer.config.use_rotary_positional_embeddings
+                    else None
+                )
+
+                # Add noise to the model input according to the noise magnitude at each timestep
+                # (this is the forward diffusion process)
+                noisy_model_input = scheduler.add_noise(model_input, noise, timesteps)
+
+                # Predict the noise residual
+                model_output = transformer(
+                    hidden_states=noisy_model_input,
+                    encoder_hidden_states=prompt_embeds,
+                    timestep=timesteps,
+                    image_rotary_emb=image_rotary_emb,
+                    return_dict=False,
+                )[0]
+                alphas_cumprod = scheduler.alphas_cumprod[timesteps]
```

Comment:
you can use `get_velocity` here 

https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_dpm_cogvideox.py#L469

Reply:
why even use dpm for training? try euler

Reply:
for now, it's to keep it consistent with the original codebase and because Cog would need a different Euler implementation than the one we have (it currently has a different implementation for DDIM and DPM too)

Reply:
but dpm isnt a training sampler, it isnt markovian. in fact using ddim and euler for training have some issues that ddpm wouldnt.